mlflow/models/utils.py

```diff
@@ -856,6 +856,73 @@ def all_ints(xs):
         )
 
 
+def _enforce_param_datatype(value: Any, dtype: DataType):
+    """
+    Enforce the value matches the data type. This is used to enforce params datatype.
+    The returned data type is a python native type.
+
+    The following type conversions are allowed:
+
+    1. int -> long, float, double
+    2. long -> float, double
+    3. float -> double
+    4. any -> datetime (try conversion)
+
+    Any other type mismatch will raise error.
+
+    Args:
+        value: parameter value
+        dtype: expected data type
+    """
+    if value is None:
+        return
+
+    if dtype == DataType.datetime:
+        try:
+            datetime_value = np.datetime64(value).item()
+            if isinstance(datetime_value, int):
+                raise MlflowException.invalid_parameter_value(
+                    f"Failed to convert value to `{dtype}`. "
+                    f"It must be convertible to datetime.date/datetime, got `{value}`"
+                )
+            return datetime_value
+        except ValueError as e:
+            raise MlflowException.invalid_parameter_value(
+                f"Failed to convert value `{value}` from type `{type(value)}` to `{dtype}`"
+            ) from e
+
+    # Note that np.isscalar(datetime.date(...)) is False
+    if not np.isscalar(value):
+        raise MlflowException.invalid_parameter_value(
+            f"Value must be a scalar for type `{dtype}`, got `{value}`"
+        )
+
+    # Always convert to python native type for params
+    if DataType.check_type(dtype, value):
+        return dtype.to_python()(value)
+
+    if (
+        (
+            DataType.check_type(DataType.integer, value)
+            and dtype in (DataType.long, DataType.float, DataType.double)
```

Comment:
nit: shall we create a mapping between original dtype and converted dtype like this?

```python
ALLOWED_TYPES = {
  DataType.integer: [DataType.long, DataType.float, DataType.double],
  DataType.long: [DataType.float, DataType.double],
  DataType.float: [DataType.double],
}
if any(DataType.check_type(original_type, value) and dtype in allowed_types) for original_type, allowed_types in ALLOWED_TYPES.items()):
 ...
```

Reply:
Make sense :)